Update the distributed throttling page to use binary publisher (#41)
wso2/docs-mg: en/docs/how-tos/rate-limiting/distributed-throttling.md

# Distributed Throttling

WSO2 API Microgateway has an in-memory mechanism by default, to handle throttling

> ([node-level throttling]({{base_path}}/how-tos/rate-limiting/adding-throttling-policies)).  

Reviewer: I've noticed if we define urls in format `{{base_path}}/<path>` it always open the link in a new tab assuming its a external link. however if we omit `{{base_path}}` and keep only the absolute path to the linked page (from site root `/`) it opens in the same tab.
```suggestion
([node-level throttling](/how-tos/rate-limiting/adding-throttling-policies)).  
```

decision is made based on the local counter within each node. If we proceed with the node local throttling in such 
environment, the API user would be allowed to send multiples of the throttling limit.I.e. if the throttling limit is set to 10, 
if we have 3 gateways in a cluster, it will allow 30 requests to pass to the backend, before all three gateways 
throttle out requests. This will put an unexpected load on the backend. To address this requirement, the API Microgateway 
supports distributed throttling where it is able to work with a central traffic management solution. In this case, 
multiple microgateways can connect with WSO2 API Manager 
([WSO2 Traffic Manager](https://apim.docs.wso2.com/en/latest/install-and-setup/setup/distributed-deployment/product-profiles/)) 
and perform rate-limiting precisely.

!!! note
    The WSO2 Traffic Manager refers to the WSO2 API Manager running in 
    [Traffic Manager Profile](https://apim.docs.wso2.com/en/latest/install-and-setup/setup/distributed-deployment/product-profiles/). 
    If you start the WSO2 API Manager without providing any profile, it runs as All in One Node (All the profiles 

>     are activated). For testing purposes, you can simply start the API Manager following quick start guide and test 

Reviewer 2: "If you start the WSO2 API Manager without providing any profile, it runs as All in One Node (All the profiles are activated). For testing purposes, you can simply start the API Manager following quick start guide and test" I suppose we can change the sentence/ change the order to make this less confusing?

Author: yes. I will make the change.

Author: updated via https://github.com/wso2/docs-mg/pull/41/commits/dae05bde11563c05e47375d5d63004a64b7ea4e4

>            Project 'petstore' is initialized successfully.

Reviewer 2: ```suggestion
```

           Project 'petstore' is initialized successfully.
           ```

2.  Now let's add the API(open API definition) to the project. Navigate to the `/petstore/api_definitions` directory. 
    Add the API definition(s) to this directory. A sample open API definition can be found 
    [here](https://github.com/wso2/product-microgateway/blob/master/samples/petstore_basic.yaml). To provide an API 

>     Level throttling tier, add the following extension to the openAPI in root level.

Reviewer 2: ```suggestion
    Level throttling tier, add the following extension to the openAPI at the root level.
```

    
    ```text tab="Sample"
    x-wso2-throttling-tier : "5PerMin"    
    ```
    
3.  The policies added in the OpenAPI definition should be deployed in the Traffic Manager. (If you are adding the
    tier "5PerMin" in openAPI definition, you need to create and deploy it in the Traffic Manager).
    The relevant documentation can be found [here](https://apim.docs.wso2.com/en/3.2.0/learn/rate-limiting/adding-new-throttling-policies/#adding-a-new-advanced-throttling-policy).

4.  Build the microgateway distribution for the project using the following command:

    ``` java tab="Format"
    micro-gw build <project_name>
    ```

    ``` java tab="Example"
    micro-gw build petstore

>     Build successful  for the project - petstore

Reviewer 2: Do we need this line for the example? 
```suggestion
```

    ```

    Once the above command is executed, An executable file ( `/petstore/target/petstore.jar` ) is created to expose the API via WSO2 API Microgateway

5.  Navigate to the `<MGW-RUNTIME-HOME>/conf/` directory and open the “micro-gw.conf" file.

6.  Enable the `enabledGlobalTMEventPublishing` property found inside the `throttlingConfig` tag. This will allow the API Microgateway to connect with the central traffic manager.

>     ``` java

Reviewer 2: ```suggestion
    ``` toml
```

     
    <table>
        <thead>
        <tr class="header">
            <th>Property</th>
            <th>Default value</th>
            <th>Description</th>
            </tr>
        </thead>
        <tbody>
            <tr class="odd">
                <td><code>receiverURL</code></td>
                <td><code>"tcp://localhost:9611"</code></td>
                <td>The URL to which the thorttle events are sent.</td>
            </tr>
            <tr class="even">
                <td><code>authURL</code></td>
                <td><code>"ssl://localhost:9711"</code></td>
                <td>The URL to which the credentials required for authentication, are sent.</td>
            </tr>
        </tbody>
    </table>

8.  If you are using API Manager 3.2.0 or later version, you need to configure the API Manager Eventhub. This step is 
    not required if you need to do only the resource level throttling or API level throttling.
    <!---TODO:@VirajSalaka Add the Event hub configuration guide URL--->
    
9.  Finally the added configurations in `microgw.conf` would look like this. 
    

>     ``` text tab="Sample Configuration"

Reviewer 2: ```suggestion
    ``` toml tab="Sample Configuration"
```

    ```

7.  In the "micro-gw.conf" file under \[throttlingConfig\]  configure the following. The purpose here is to configure 
    the message broker.

    <table>
        <thead>
            <tr class="header">
                <th>Property</th>
                <th>Default value</th>
                <th>Description</th>
            </tr>
        </thead>
        <tbody>
            <tr class="odd">
                <td><code>jmsConnectionProviderUrl</code></td>
                <td><code>"amqp://admin:admin@carbon/carbon?brokerlist='tcp://localhost:5672'"</code></td>
                <td>The message broker connection URL of WSO2 API/Traffic Manager</td>
            </tr>
            <tr class="even">
                <td><code>jmsConnectionUsername</code></td>
                <td><code>""</code></td>
                <td>The user name used to establish  the message broker connection</td>
            </tr>
            <tr class="odd">
                <td><code>jmsConnectionPassword</code></td>
                <td><code>""</code></td>
                <td>The password used to establish  the message broker connection</td>
            </tr>
        </tbody>
    </table>
    

>     The message broker connection URL. For e.g. a [WSO2 API instance can be used as the Traffic Manager](https://docs.wso2.com/display/AM260/Product+Profiles). 

Reviewer: Shall we point to 320 apim docs?

Reviewer: ```suggestion
    The message broker connection URL. For e.g. a [WSO2 API instance can be used as the Traffic Manager](https://apim.docs.wso2.com/en/latest/install-and-setup/setup/distributed-deployment/product-profiles/). 
```

    In the "micro-gw.conf" file under \[throttlingConfig.binary\], we should list down all the configurations related to
    event publishing.                   
        
    <table>
        <thead>
        <tr class="header">
            <th>Property</th>
            <th>Default value</th>
            <th>Description</th>
            </tr>
        </thead>
        <tbody>
            <tr class="odd">
                <td><code>enabled</code></td>
                <td><code>true</code></td>
                <td>Enable the binary event publisher. If it is `false`, http event publisher will be enabled</td>
            </tr>
            <tr class="even">
                <td><code>username</code></td>
                <td><code>"admin"</code></td>
                <td>The user name used for authentication prior to publishing events via binary publisher</td>
            </tr>
            <tr class="odd">
                <td><code>password</code></td>
                <td><code>"admin"</code></td>
                <td>The password used for authentication prior to publishing events via binary publisher</td>
            </tr>
        </tbody>
    </table>
    

>     The binary receiver URL/URLs need to be added as an Array using the key \[\[throttlingConfig.binary.URLGroup\]\]. 

Reviewer: ```suggestion
    The binary receiver URL(s) need to be added as an Array using the key \[\[throttlingConfig.binary.URLGroup\]\]. 
```
